cmd/serve/option.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serve
+
+import (
+	"net"
+	"os"
+	"strings"
+
+	"github.com/cilium/hubble/pkg/api"
+	v1 "github.com/cilium/hubble/pkg/api/v1"
+	"github.com/cilium/hubble/pkg/server"
+	"golang.org/x/sys/unix"
+	"google.golang.org/grpc/health"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+)
+
+// DefaultOptions is the reference point for default values.
+var DefaultOptions = Options{
+	Listeners: make(map[string]net.Listener),
+}
+
+// Options stores all the configuration values for the hubble server.
+type Options struct {
+	Listeners       map[string]net.Listener
+	HealthService   *health.Server
+	ObserverService *server.GRPCServer
+}
+
+// Option customizes then configuration of the hubble server.
+type Option func(o *Options) error
+
+// WithTCPListeners configures TCP listeners with the given set of addresses.
+func WithTCPListeners(addresses []string) Option {
+	return func(o *Options) error {
+		var err error
+		defer func() {
+			if err != nil {
+				for _, l := range o.Listeners {
+					l.Close()
+				}
+			}
+		}()
+		for _, address := range addresses {
+			if address == "" || strings.HasPrefix(address, "unix://") {
```

Comment:
I think this check for `unix://` should be removed here.

Reply:
The problem is that the listen addresses of the command is a slice of potentially mixed unix domain sockets and tcp addresses. This check prevents trying to configure a unix domain socket as a tcp socket. Ultimately though, I agree, it should be cleaned up one way or the other.

Reply:
Ah, I see. Could we add a `WithListeners(addresses []string)` option which internally calls `WithTCPListener(string)` or `WithUnixSocketListener(string)`?

But we can do that in a follow-up PR.

Reply:
Done.

Reply:
Thanks! :100: